### PR TITLE
funcapp: Add support for deploying flex-consumption function apps

### PR DESCRIPTION
**Pull request by weikanglim**
Add support for deploying flex-consumption function apps.

To do this, we call into the new `api/publish` endpoint that works only for flex-consumption function apps.

Addresses #3658

### CODE REVIEW

File: cli/azd/pkg/tools/azcli/function_app.go [hunk 1]
```diff
@@ -44,6 +49,37 @@ func (cli *azCli) DeployFunctionAppUsingZipFile(
...
+	}
+
+	plansClient, err := armappservice.NewPlansClient(subscriptionId, cred, cli.armClientOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	planId := *app.Properties.ServerFarmID
+	sep := strings.LastIndexByte(planId, '/')
+	planName := planId[sep+1:]
```
Comment (vhvb1989) [inline cli/azd/pkg/tools/azcli/function_app.go:64]: should we check sep != -1 as an error  here?

**Review by vhvb1989 — APPROVED** [review]
Added some feedback, otherwise LGTM

File: cli/azd/pkg/azsdk/funcapp_host_client.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+	credential azcore.TokenCredential,
+	options *arm.ClientOptions,
+) (*FuncAppHostClient, error) {
+	funcAppHostOptions := &arm.ClientOptions{}
+	if options != nil {
+		optionsCopy := *options
+		funcAppHostOptions = &optionsCopy
+	}
+
+	options.DisableRPRegistration = true
```
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/pkg/azsdk/funcapp_host_client.go:36]: Yes it should, missed this in the latest revision. Thanks for the catch!

File: cli/azd/pkg/azsdk/funcapp_host_client.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+const (
+	PublishStatusCancelled      int = -1
+	PublishStatusPending        int = 0
+	PublishStatusBuilding       int = 1
+	PublishStatusDeploying      int = 2
+	PublishStatusFailed         int = 3
+	PublishStatusSuccess        int = 4
+	PublishStatusConflict       int = 5
+	PublishStatusPartialSuccess int = 6
```
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/pkg/azsdk/funcapp_host_client.go:175]: I think we should, we export the `PublishResponse` and we should expose the status as well. I created a type alias for it.

File: cli/azd/pkg/tools/azcli/function_app.go (hunk 1 above)
Comment (weikanglim) [reply cli/azd/pkg/tools/azcli/function_app.go:64]: Makes sense

File: cli/azd/pkg/azsdk/funcapp_host_client.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+type publishPollingHandler struct {
+	pipeline runtime.Pipeline
+	location string
+
+	lastResponse *PublishResponse
+	// final result
+	result *PublishResponse
+}
+
+func newPublishPollingHandler(
```
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/pkg/azsdk/funcapp_host_client.go:188]: You are definitely right about the complexity here. The minor things we're getting from the SDK:

1. Polling retry delay is handled in `PollUntilDone()`. If the API returns Success, but with a retry header present, the delay is automatically added to the Polling delay.
2. Slightly more logging for loop polling
3. "For-loop as a component" -- A `Begin*` that returns a `Poller` that decide when to poll for the LRO.

And the downside:

Ceremony for for-loop abstraction. It does make it trickier when you're thinking about adding more things like progress logging in the future. The handler really is a "for-loop abstraction" that isn't as easy to reason about, and sometimes thinking about what to pass as a variable doesn't come naturally.

I don't see us using 2. and 3. anytime soon, and I do value having a simpler for loop to work with. I think it's worth experimenting with this in the codebase and see how it feels over time. So I'll adopt the suggestion here and let's go with it. To overcome 1., I've added the `RetryAfter` implementation in `httputil` that works the same as the one in the azure-sdk-for-go's `internal/shared` package.

**Review by wbreza — APPROVED** [review]
Added a couple minor comments but looks good overall.

File: cli/azd/pkg/azsdk/funcapp_host_client.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+	// or activating the virtual environment for Python apps.
+	RemoteBuild bool
+}
+
+// Publish deploys an application zip file to the function app host.
+// This is currently only supported for flexconsumption plans.
+func (c *FuncAppHostClient) Publish(
+	ctx context.Context,
+	zipFile io.ReadSeeker,
+	options *PublishOptions) (PublishResponse, error) {
```
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/pkg/azsdk/funcapp_host_client.go:63]: Opting to return empty struct here and be more inline with azure-sdk-for-go.

File: cli/azd/pkg/tools/azcli/function_app.go [hunk 6]
```diff
@@ -44,6 +49,41 @@ func (cli *azCli) DeployFunctionAppUsingZipFile(
...
+		return nil, err
+	}
+
+	planId := *app.Properties.ServerFarmID
+	sep := strings.LastIndexByte(planId, '/')
+	if sep == -1 {
+		return nil, fmt.Errorf("unexpected ServerFarmID %s", planId)
+	}
+
+	planName := planId[sep+1:]
```
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/pkg/tools/azcli/function_app.go:68]: We could use `path.Base` (not `filepath` which handles OS paths, including forward and backward slashes). I think if we did, we'd have to handle both `.` and `/` as possible return values, and there wouldn't be an easy way to tell that there were no slashes (`path.Base` would have returned the input without modification).

File: cli/azd/pkg/azsdk/funcapp_host_client.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+		delay := polLDelay
+		if retryAfter := httputil.RetryAfter(response); retryAfter > 0 {
+			delay = polLDelay
+		}
+
+		select {
+		case <-ctx.Done():
+			return PublishResponse{}, ctx.Err()
+		case <-time.After(delay):
+		}
```
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/pkg/azsdk/funcapp_host_client.go:179]: Slightly prefer the current version (a poll is a full execution of the loop), leaving it as-is.

File: cli/azd/pkg/azsdk/funcapp_host_client.go [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+	if deploymentId == "" {
+		return PublishResponse{}, fmt.Errorf("missing deployment id")
+	}
+
+	location := fmt.Sprintf("https://%s/api/deployments/%s", c.hostName, deploymentId)
+	return c.waitForDeployment(ctx, location)
+}
+
+func (c *FuncAppHostClient) waitForDeployment(ctx context.Context, location string) (PublishResponse, error) {
```
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/pkg/azsdk/funcapp_host_client.go:118]: There isn't a built-in poller for this specific interaction. With respect to using azure-sdk-for-go's poller implementation for generic LROs, I don't see us taking advantage of it anytime soon, see previous comment [here](https://github.com/Azure/azure-dev/pull/4003#discussion_r1636959552).

**Comment by pamelafox** [comment]
I tried out this build and got this error:

```
Deploying services (azd deploy)

  (x) Failed: Deploying service api

ERROR: error executing step command 'deploy --all': failed deploying service 'api': publishing zip file: deployment failed: Response status code does not indicate success: 404 (The specified container does not exist.).
```

My code is in:
https://github.com/pamelafox/simple-flask-api-azure-function/pull/15/files

There's a 100% chance that I bungled something, but am not sure what.

**Comment by weikanglim** [comment]
@pamelafox The storage container for storing the deployment artifacts isn't created automatically by the azure functions host, so it needs to be pre-created in bicep. See the example in [here](https://github.com/weikanglim/todo-python-mongo-swa-func/commit/75ac6a252d19482210915b40cdfabf965566990f#diff-7ef659fc9cf6968e718894d300490b14ea7a52091e7d4bcffae3a5029ac721d4).

**Comment by pamelafox** [comment]
@weikanglim Thank you! It seems to be working now, with those changes. By the way, that Bicep still has the storage connection string, but that shouldn't be needed anymore. I removed it and had only this:

AzureWebJobsStorage__accountName: storageAccount.name

**Comment by weikanglim** [comment]
@pamelafox Ah yes, that is indeed the whole point of removing the need to directly grab the access keys! You're right that we can replace `AzureWebJobsStorage` with `AzureWebJobsStorage__accountName`.

**Comment by sinedied** [comment]
Sent a PR for supporting VNET and storage managed identity in Functions: https://github.com/Azure/azure-dev/pull/4026

**Pull request merged**